src/osd/modules/netdev/vmnet.cpp

```diff
@@ -0,0 +1,280 @@
+
+	if (interface_status == VMNET_SUCCESS) {
+		long buffer_size = (m_vmnet_packet_size * 2 + 1023) & ~1023;
+		m_buffer = (uint8_t *)malloc(buffer_size);
+		// TODO: investigate using vmnet_interface_set_event_callback to set up a callback when data is available.
+	} else {
+		if (m_interface) {
+			vmnet_stop_interface(m_interface, q, ^(vmnet_return_t status){
+				dispatch_semaphore_signal(sem);
+			});
+			dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
+			m_interface = NULL;
+		}
+		osd_printf_error("vmnet_start_interface failed: %d\n", interface_status);
+	}
+
+	dispatch_release(sem);
+	xpc_release(dict);
+}
+
+netdev_vmnet::~netdev_vmnet() {
+	vmnet_return_t st;
+	dispatch_queue_t q;
+	dispatch_semaphore_t sem;
+
+	if (m_interface) {
+		sem = dispatch_semaphore_create(0);
+		q = dispatch_get_global_queue(QOS_CLASS_UTILITY, 0);
+
+		st = vmnet_stop_interface(m_interface, q, ^(vmnet_return_t status){
+			dispatch_semaphore_signal(sem);
+		});
+		if (st == VMNET_SUCCESS) {
+			dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
+		}
+		dispatch_release(sem);
+	}
+	free(m_buffer);
+}
+
+void netdev_vmnet::set_mac(const char *mac)
+{
+	memcpy(m_mac, mac, 6);
+}
+
+
+int netdev_vmnet::send(uint8_t *buf, int len)
+{
+	int count = 1;
+	vmnet_return_t st;
+	struct vmpktdesc v;
+	struct iovec iov;
+
+	if (!m_interface) return 0;
+
+	if (len > m_vmnet_packet_size) {
+		osd_printf_error("vmnet: packet too big (%d)\n", len);
+		return 0;
+	}
+
+	if (memcmp(m_mac, m_vmnet_mac, 6) != 0) {
+		// TODO: do we need 2 buffers, in case read recv buffer still in use?
+		memcpy(m_buffer, buf, len);
+		vmnet_common::fix_outgoing_packet(m_buffer, len, m_vmnet_mac, m_mac);
+		buf = m_buffer;
+	}
+
+	iov.iov_base = buf;
+	iov.iov_len = len;
+
+	v.vm_pkt_size = len;
+	v.vm_pkt_iov = &iov;
+	v.vm_pkt_iovcnt = 1;
+	v.vm_flags = 0;
+
+	st = vmnet_write(m_interface, &v, &count);
+
+	if (st != VMNET_SUCCESS) {
+		osd_printf_error("vmnet_write: %d\n", st);
+		return 0;
+	} else {
+		osd_printf_error("vmnet_write sent %d OK\n", len);
+	}
+
+	return count;
+}
+
+int netdev_vmnet::recv_dev(uint8_t **buf) {
+
+	int count = 1;
+	vmnet_return_t st;
+	struct vmpktdesc v;
+	struct iovec iov[2];
+
+	if (!m_interface) return 0;
+
+	for(;;) {
+		int type;
+
+		iov[0].iov_base = m_buffer;
+		iov[0].iov_len = m_vmnet_packet_size;
+
+		v.vm_pkt_size = m_vmnet_packet_size;
+		v.vm_pkt_iov = iov;
+		v.vm_pkt_iovcnt = 1;
+		v.vm_flags = 0;
+
+		count = 1;
+		st = vmnet_read(m_interface, &v, &count);
+		if (st != VMNET_SUCCESS) {
+			osd_printf_error("vmnet_read: %d\n", st);
+			return 0;
+		}
+
+		if (count < 1) break;
+
+		/* todo -- skip multicast messages based on flag? */
+		type = classify_mac(m_buffer);
+		if (type == 2) continue; /* multicast */
+		break;
+	}
+
+	if (count < 1) return 0;
+
+	osd_printf_verbose("MAC is %02x:%02x:%02x:%02x:%02x:%02x\n", m_mac[0], m_mac[1], m_mac[2], m_mac[3], m_mac[4], m_mac[5]);
+	if (memcmp(m_mac, m_vmnet_mac, 6) != 0) {
```

Comment:
This will get really spammy if there’s significant traffic.  Also, isn’t the only place the contents of `m_mac` can change in `set_mac`?  What’s the point of logging it on every received packet?